pkg/syncer/syncer.go

```diff
@@ -120,37 +159,57 @@ func (c *Controller) delete(ctx context.Context, gvr schema.GroupVersionResource
 }
 
 func (c *Controller) upsert(ctx context.Context, gvr schema.GroupVersionResource, namespace string, unstrob *unstructured.Unstructured) error {
+
 	client := c.getClient(gvr, namespace)
 
 	// Attempt to update the object; if the object isn't found, create it.
-	unstrob.SetResourceVersion("")
+
 	unstrob.SetUID("")
-	if _, err := client.Update(ctx, unstrob, metav1.UpdateOptions{}); err != nil {
-		if k8serrors.IsNotFound(err) {
-			if _, err := client.Create(ctx, unstrob, metav1.CreateOptions{}); err != nil {
-				return err
-			}
-		} else {
+	existing, err := client.Get(ctx, unstrob.GetName(), metav1.GetOptions{})
+	if err == nil {
+		klog.Infof("Object %s/%s already exists: update it", gvr.Resource, unstrob.GetName())
+
+		unstrob.SetResourceVersion(existing.GetResourceVersion())
+		if _, err = client.Update(ctx, unstrob, metav1.UpdateOptions{}); err != nil {
+			klog.Error(err)
+			return err
+		}
+	} else if k8serrors.IsNotFound(err) {
+		klog.Infof("Object %s/%s doesn't already exist: create it", gvr.Resource, unstrob.GetName())
+		unstrob.SetResourceVersion("")
+		if _, err = client.Create(ctx, unstrob, metav1.CreateOptions{}); err != nil {
+			klog.Error(err)
 			return err
 		}
+	} else {
+		klog.Error(err)
+		return err
+	}
+
+	klog.Infof("Update the status of object %s/%s", gvr.Resource, unstrob.GetName())
+	_, err = client.UpdateStatus(ctx, unstrob, metav1.UpdateOptions{})
```

Comment:
Style nit: I tend to prefer the `if _, err := client.UpdateStatus(...); err != nil { ... }` form for things like this, since the value of `err` won't escape the scope of this small block.

Then you can either `return err` inside the block, or `return nil` outside of it.

Or, as before, just get the value of `err`, then return it (whether nil or not). This change seems to be in service of logging when `UpdateStatus` fails, which will already happen when `upsert` fails, when `process` fails, when `handleErr` sees it.

Reply:
I'll still `klog` here to have the precise error location in the log message, but will refactor it in your preferred form :-)

Reply:
as much as possible :-)